### DISCUSSION TITLE
Dexie liveQuery()'s querier callback did'nt await all of its spawned promises.

### DISCUSSION THREAD

**Discussion by dusty-phillips** [dexie/Dexie.js#1839] [Q&A]
I'm experimenting with using Dexie with [solidjs](https://www.solidjs.com), using a few dozen lines of code borrowed from https://github.com/faassen/solid-dexie/blob/main/src/solid-dexie.ts. I've encountered what I'm going to guess is a dumb question.

The connecting logic is essentially just hooking up `liveQuery` to solidjs's reactivity bindings. It works as designed with simple queries that just proxy straight through to a dexie Table method. But when I try to do something more complicated inside an async querier, I get this warning on the console:

```
Dexie liveQuery()'s querier callback did'nt await all of its spawned promises. Querier source: () => getOutlineItemAndSceneText(selectedOutlineItemId())
```

So, two things there: One, there's a typo on `did'nt`. Two: I can't find any reference to what this means. Who is the culprit? Is it that my querier function isn't awaiting something? Or is it that the borrowed solid-dexie code isn't querying anything? Or is it a bug in dexie (unlikely).

I'm using Dexie version `4.0.1-beta.4`, and if it's important, it's connected to `dexie-cloud-addon` `4.0.1-beta.53`.

The querier function in question looks like this:

```
export const getSelectedOutlineItemId = async (draft: Draft) => {
  if (draft.selectedOutlineItemId) {
    if (await db.outlineItems.get(draft.selectedOutlineItemId)) {
      return draft.selectedOutlineItemId
    }
  }

  const outlineItemsForDraft = await db.outlineItems
    .where('draftid')
    .equals(draft.id)
    .sortBy('order')

  return outlineItemsForDraft[0].id
}
```

~I'm pretty sure the error would manifest with any async querier function.~ (I tested this assertion and it does not always repro. Still trying to find the connection)

I'm connecting it to solid-dexie using:

```
  const selectedOutlineItemId = createDexieSignalQuery(() =>
    getSelectedOutlineItemId(routeData.selectedDraft()),
  )
```

I'm using a slightly hacked version of the solid-dexie code that looks like this:

```
export function createDexieSignalQuery<T>(
  querier: () => NotArray<T> | Promise<NotArray<T>>,
): Accessor<T | undefined> {
  const get = createMemo(() => from<T>(liveQuery(querier)))
  return () => get()()
}
```

I'm pretty confident this is going to be a "dumb question" kinda answer, but I'm not sure where to find it on my own.

**Comment by dusty-phillips** [comment]
Realized this warning was added to dexie 5 days ago to resolve https://github.com/dexie/Dexie.js/issues/1831 so I’m wondering if maybe I’ve encountered an edge case. I’m not spawning any promises inside the queries that I’m not awaiting, as far as I can tell. Is the check too aggressive?

**Comment by dfahlander** [comment]
Thanks for the report! I should probably try solid-dexie and see why it warns on this. Looking at the code you pasted I don't see any un-awaited promise so it might be a false positive. This warning was meant to detect queriers that did things like:

```ts
  someArray.forEach(async () => ....) // Problematic - spawns promises without awaiting
```

instead of the correctly awaiting everything:

```ts
  await Promise.all(someArray.map(async () => ...)) // correct!
```

or possibly:

```ts
  for (const item of someArray) {
    await ... // correct!
  }
```

...which could lead to returning unfinished results and also possible performance degradation.

I'll try see if I get the same with solid-dexie and find the reason for it. If you have a repro of it, please share! Otherwise I'll just try with solid-dexie and see if I get the same warning.

**Reply by dusty-phillips** [reply]
My current guess is that the solid compiler (which I don't know anything about) is abstracting something away that is causing the false positive. I don't have a simple repro handy right now, but I'll put something together later this week if you don't get to it.

**Reply by dusty-phillips** [reply]
Heads up that I tried to make a minimal repro and wasn't able to, so my new current assumption is that something in my real code is actually broken. Will keep you posted.

**Reply by dusty-phillips** [reply]
Circling back here, though not with a satisfying conclusion: I eventually decided that solid doesn't work for the way I want to think about components, so I didn't make any progress on identifying the problem.

**Reply by dfahlander** [reply]
ok I see. I think though that there were issues in dexie giving that warning in false, so should not be solid to blame. Latest dexie (4.0.1-beta.5) does not warn about it anyhow.
